packages/mcfunction/src/parser/macro.ts

```diff
@@ -0,0 +1,161 @@
+import * as core from '@spyglassmc/core'
+import { localeQuote, localize } from '@spyglassmc/locales'
+import type { MacroChildNode, MacroKeyNode, MacroNode } from '../node/index.js'
+
+/**
+ * Parse a macro line.
+ */
+export function macro(): core.Parser<MacroNode> {
+	return (
+		src: core.Source,
+		ctx: core.ParserContext,
+	): core.Result<MacroNode> => {
+		const start = src.cursor
+		const children: MacroChildNode[] = []
+
+		// Handle the starting '$'
+		const sign: MacroChildNode = {
+			type: 'mcfunction:macro_child/sign',
+			range: core.Range.create(start, start + 1),
+			value: '$',
+		}
+		children.push(sign)
+		src.skip()
+
+		// Handle the rest of the line
+
+		let beginning = src.cursor
+		let nextChunk = src.readUntil('$', '\r', '\n')
+		do { // Do-while because we always want it to work the first time
+			let wasMacro = false
+			const child: MacroChildNode = parseMacroChild(
+				src,
+				ctx,
+				nextChunk,
+				beginning,
+			)
+			if (child.type === 'mcfunction:macro_child/macro') {
+				wasMacro = true
+			}
+			// Handle a non-macro $ in the middle of the line
+			let subMacro: MacroChildNode | undefined
+			if (
+				child.type === 'mcfunction:macro_child/other' &&
+				child.value?.indexOf('$') !== -1
+			) {
+				const parts: String[] = child.value?.split('$') ?? []
+				let other = ''
+				for (const part of parts) {
+					if (part.charAt(0) == '(') { // We've found our macro key!
+						subMacro = parseMacroChild(
+							src,
+							ctx,
+							'$' + part,
+							beginning + other.length - 1,
+						)
+						child.range.end -= part.length + 1
+					} else { // Keep it in the normal string
+						other += part + '$'
+					}
+				}
+				wasMacro = true
+				child.value = other.substring(0, other.length - 1) // Remove the last '$'
+			}
+			children.push(child)
+			if (subMacro) { // Sub-macro added here because colorizer
+				children.push(subMacro)
+			}
+			// Prepare for the next block
+			beginning = src.cursor
+			if (wasMacro) {
+				nextChunk = src.readUntil('\r', '\n', '$')
+			} else {
+				nextChunk = src.readUntil('\r', '\n', ')')
```

Comment:
i see a handful of `\r` and `\n` checks, you can probably just wrap the whole parser in `stopBefore('\r', '\n')` and remove these explicit checks

we do that for the `argument` parser:

https://github.com/SpyglassMC/Spyglass/blob/2b74e1ff633f56db439c6084d1bb72c16f536a88/packages/java-edition/src/mcfunction/parser/argument.ts#L93-L98